// grunt-lesslint/tasks/lib/less-parser.js
(function() {
  var LessParser, Parser, defaultLessOptions, path, _;

  path = require('path');

  _ = require('lodash');

  Parser = require('less').Parser;

  defaultLessOptions = {
    cleancss: false,
    compress: false,
    dumpLineNumbers: 'comments',
    optimization: null,
    syncImport: true
  };

  module.exports = LessParser = (function() {
    function LessParser(fileName, opts) {
      var paths;
      opts = _.defaults(opts || {}, defaultLessOptions);
      paths = [path.dirname(path.resolve(fileName))];
      if (opts.less && opts.less.paths) {
        paths = paths.concat(opts.less.paths);
      }
      this.opts = _.extend({
        filename: path.resolve(fileName),
        paths: paths,
        sourceMaps: true
      }, opts);
      this.parser = new Parser(this.opts);
    }

    LessParser.prototype.parse = function(less, callback) {
      var err;
      try {
        return this.parser.parse(less, callback);
      } catch (_error) {
        err = _error;
        return callback(err);
      }
    };

    return LessParser;

  })();

}).call(this);
